[FSDP][optim_state_dict] Print out more useful error message for optim_state_dict

fegin opened the pull request:
Summary: Print out more useful error message for optim_state_dict

Test Plan: CI

Reviewed By: wz337

Differential Revision: D43556073

rohan-varma commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1534,7 +1539,7 @@ def module_fn(module, prefix, fqn_to_param_info):
             fqn = clean_tensor_name(prefix + local_fqn)
             if fqn in fqn_to_param_info:
-                assert fqn_to_param_info[fqn].flat_param == flat_param
+                assert fqn_to_param_info[fqn].flat_param is flat_param, fqn
```
add some details about how this error can happen to the msg?

rohan-varma reviewed (approved):
LGTM

wz337 reviewed (approved):
Stamped in Phabricator.

fegin committed:
[FSDP][optim_state_dict] Print out more useful error message for optim_state_dict (#96860)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/96860

Print out more useful error message for optim_state_dict

Test Plan: CI

Reviewed By: wz337

Differential Revision: D43556073

fbshipit-source-id: 5e488018bf1a47412fd8cd6fd1b2844a095546a2
